tv/tv.cpp

```diff
@@ -591,56 +605,147 @@ bool do_skip(const llvm::StringRef &pass0) {
                 [&](auto skip) { return pass == skip; });
 }
 
+struct TVNewPass : public llvm::PassInfoMixin<TVNewPass> {
+  static bool skip_tv;
+  static string pass_name;
+  // The number of TVNewPass created.
+  static unsigned num_tv_pass_created;
+  bool print_pass_name = false;
+
+
+  llvm::PreservedAnalyses run(llvm::Module &M,
+                              llvm::ModuleAnalysisManager &AM) {
+    auto &FAM = AM.getResult<llvm::FunctionAnalysisManagerModuleProxy>(M)
+                  .getManager();
+    auto get_TLI = [&FAM](llvm::Function &F) {
+      return &FAM.getResult<llvm::TargetLibraryAnalysis>(F);
+    };
+    run(M, get_TLI);
+    return llvm::PreservedAnalyses::all();
+  }
+
+  void run(llvm::Module &M,
+           function<llvm::TargetLibraryInfo*(llvm::Function&)> get_TLI) {
+    // Check TVFinalizePass::finalized again because TVNewPass can be
+    // called after TVFinalizePass registered by
+    // registerOptimizerLastEPCallback.
+    // It can happen when a default pipeline (O3, O2, ...) is used.
+    if (TVFinalizePass::finalized)
+      return;
+
+    static int count = 0;
+    if (!out) {
+      // TVInitPass is not called yet.
+      // This can happen at very early passes, such as
+      // ForceFunctionAttrsPass.
+      skip_tv = false;
+      return;
+    }
+
+    count++;
+    if (print_pass_name) {
+      // print_pass_name is set only when running clang tv
+      *out << "-- " << count << ". " << pass_name
+           << (skip_tv ? " : Skipping\n" : "\n");
+    }
+
+    TVPass tv;
+    tv.skip_verify = skip_tv;
+
+    for (auto &F: M) {
+      tv.TLI_override = get_TLI(F);
+
+      // If skip_pass is true, this updates fns map only.
+      tv.runOnFunction(F);
```

Comment:
I think we should change the old PM pass to module pass as well. It will be needed for other parallelization strategies and should simplify the code here a bit.

Reply:
The diff becomes pretty large. May I make this change in a separate PR?

Reply:
sounds good!